### ISSUE TITLE
Spotify provider |  Login failed since 2.1.3

### ISSUE THREAD

**Issue by eekamouze** [music-assistant/hass-music-assistant#2754]
### What version of Music Assistant has the issue?

2.1.3

### What version of the Home Assistant Integration have you got installed?

n/a

### Have you tried everything in the Troubleshooting FAQ and reviewed the Open and Closed Issues and Discussions to resolve this yourself?

- [X] Yes

### The problem

I received a notification from spotify yesterday night telling me that my password has been reseted due to abnormal activity.
After changing my password today I realized that I cannot login anymore within MASS spotify provider. I did the upgrade to 2.1.3 24h ago and I suspect the abnormal activity comes from MASS. I rolled back to a previous version (2.0.7) and that fixed the problem

### How to reproduce

Upgrade to 2.1.3 (latest) 

### Music Providers

Spotify

### Player Providers

n/a

### What version of Home Assistant Core are your running

2024.8.1

### What type of installation are you running?

Home Assistant Container

### On what type of hardware are you running?

Raspberry Pi

**Comment by CrallH** [comment]
i have the same issue.

**Comment by Rep-Hueman** [comment]
i have the same issue

**Comment by burghs** [comment]
The issue is on Spotify side, looks like Username and Password for authentication is changing on their end. MASS beta6 release works with the new authentication method so I would imagine that the stable release will follow shortly.

**Comment by Rep-Hueman** [comment]
Better to use the login process as HA uses. Don't use username and password but a Spotify APP - Client ID with a Client secret

**Comment by DanielHabenicht** [comment]
Duplicate of https://github.com/music-assistant/hass-music-assistant/issues/2749

@OzGav I tried the beta6 and must say that the process is a little bit confusing for a normal user. Would it be possible to change the warning tab to a modal which opens on pressing the "Authenticate" button and than guides a user through authenticating instead?

**Comment by marcelveldt** [comment]
> Better to use the login process as HA uses. Don't use username and password but a Spotify APP - Client ID with a Client secret

That is not entirely that simple as MA also needs to playback actual audio and HA can only control the regular web api.

We have indeed implemented a new auth method (using spotify connect) to replace the deprecated username/password login and testing it in beta. If reports from the beta today are fine, we'll bump stable as well to have this fix.

**Comment by marcelveldt** [comment]
> This is a duplicate of #2749
> 
> @OzGav I tried the beta6 and must say that the process is a little bit confusing for a normal user. Would it be possible to change the warning tab to a modal which opens on pressing the "Authenticate" button and than guides a user through authenticating instead?

No not possible at the moment. Just read the text and you'll be fine.

**Comment by DanielHabenicht** [comment]
> > This is a duplicate of #2749
> > @OzGav I tried the beta6 and must say that the process is a little bit confusing for a normal user. Would it be possible to change the warning tab to a modal which opens on pressing the "Authenticate" button and than guides a user through authenticating instead?
> 
> No not possible at the moment. Just read the text and you'll be fine.

Alright, apart from that I can report that it works for me.

**Comment by cornebistouille** [comment]
i have the same problem with one my account (my wife's one that i shared with my premiuum account) 
i a new to this... some guys above talk about a solution, but i don't understand it...

**Comment by mdepedrof** [comment]
Same here. Premium spotify account and i cant login. I use docker with latest tag on image

**Comment by eamonnsullivan** [comment]
I'm seeing the same issue and I am already using beta 6 of the server. I got the password-reset email, changed my password, but Music Assistant is just spinning and never shows me anything else when I try to reauthenticate there.

**Comment by erkr** [comment]
@eamonnsullivan  did you open music assistant as playback destination on you Spotify app after pressing the authorize button on MA? That needs to be once to complete authentication

**Comment by eekamouze** [comment]
Worked for me after upgrading to 2.1.4 and selecting MUSIC_ASSISTANT as the playback device on my Spotify app. Also I managed to add Spotify Client ID.  Thanks.

**Comment by eamonnsullivan** [comment]
> @eamonnsullivan did you open music assistant as playback destination on you Spotify app after pressing the authorize button on MA? That needs to be done once to complete authentication! After a successful logon, it will take some time to sync all data. Especially without an own client id
> 
> I successfully re-authored two Spotify accounts, with their own client id's

Thanks. I'll have to wait until I'm home, then. I was trying to do this via my Home Assistant Cloud connection from the office, which obviously isn't going to work. Could this be documented better, maybe? It was not obvious.

**Comment by marcelveldt** [comment]
Let's leave this one open for others to quickly spot it and close once the dust settles down.

**Comment by firefrei** [comment]
Does your Spotify App needs to be on the same L2 network than music assistant?
If this is the case, probably this information also helps others (as I couldn't find it in the docs so far). 

I'm running 2.1.4 and clicked the `Authenticate Spotify` button, but the wheel keeps spinning. In my Spotify app I do not see any `Music_Assistant` device (Well, how should the app find the MA instance if I do not/cannot enter anything like an ID :-)).
Is there any other auth. option / enhancement planned for future release to support this, or do I need to move MA to my user wifi network 'just' for the auth. process?

Many thanks to the MA team for providing a fix in such a short time!

**Comment by OzGav** [comment]
Yes. And see three posts above yours

**Comment by diogotcorreia** [comment]
@firefrei I had the same problem. I ended up downloading the librespot binary (https://github.com/music-assistant/server/blob/1d49e4061d50299195e195ff39e121423647a44c/music_assistant/server/providers/spotify/bin/librespot-linux-x86_64) and running it (see below) on my desktop (which is also running a Spotify official client).
The device then showed up on my PC in the Spotify app.
Then, copy the `./data-dir/credentials.json` file to the `spotify` (or `spotify-setup`, whichever exists) directory in the MA data directory (`/data` on the container).

To run librespot use the command: `./librespot-linux-x86_64 -c ./data-dir -a -n MUSIC_ASSISTANT` (you might need to create `./data-dir` manually)

Note: you might need to reload the spotify provider after moving the file.

**Comment by marcelveldt** [comment]
Folks, just follow our documentation and make sure MA sits on the same L2 network as your devices (and Spotify app in this case) without firewalls so that hacks like this are not needed.

**Comment by Intellium** [comment]
> Does your Spotify App needs to be on the same L2 network than music assistant? If this is the case, probably this information also helps others (as I couldn't find it in the docs so far).
> 
> I'm running 2.1.4 and clicked the `Authenticate Spotify` button, but the wheel keeps spinning. In my Spotify app I do not see any `Music_Assistant` device (Well, how should the app find the MA instance if I do not/cannot enter anything like an ID :-)). Is there any other auth. option / enhancement planned for future release to support this, or do I need to move MA to my user wifi network 'just' for the auth. process (my MA instance is running behind a reverse proxy on my server, so it is not directly attached to any user network)?
> 
> Many thanks to the MA team for providing a fix in such a short time!

I'm having the same issue. But even using the app in the same L2 as where music assistant is running, there is no Music_Assistant player. Even though I can see every other player.

**Comment by Intellium** [comment]
> Folks, just follow our documentation and make sure MA sits on the same L2 network as your devices (and Spotify app in this case) without firewalls so that hacks like this are not needed.

Am I missing something?
The docs here https://music-assistant.io/music-providers/spotify/ still mention loging in with username and password. Nothing is mentioned about the new authentication flow?

**Comment by eamonnsullivan** [comment]
> Am I missing something?

No, it's still the old docs. However, there's [PR pending](https://github.com/music-assistant/music-assistant.github.io/pull/85). It's mentioned in a comment above.

**Comment by Intellium** [comment]
> @firefrei I had the same problem. I ended up downloading the librespot binary (https://github.com/music-assistant/server/blob/1d49e4061d50299195e195ff39e121423647a44c/music_assistant/server/providers/spotify/bin/librespot-linux-x86_64) and running it (see below) on my desktop (which is also running a Spotify official client). The device then showed up on my PC in the Spotify app. Then, copy the `./data-dir/credentials.json` file to the `spotify` (or `spotify-setup`, whichever exists) directory in the MA data directory (`/data` on the container).
> 
> To run librespot use the command: `./librespot-linux-x86_64 -c ./data-dir -a -n MUSIC_ASSISTANT` (you might need to create `./data-dir` manually)
> 
> Note: you might need to reload the spotify provider after moving the file.

I had to do the same thing to get it working again. 
My music assistant is running in k8s. Allthough there is a L2 network available in the pod (via multus) on which it should be able to reach everything, I'm thinking the authentication flow doesn't use the setting to override IP ?

**Comment by robin-thoni** [comment]
> My music assistant is running in k8s

That's the reason I hate L2 discovery. I would also like to move to this kind of setup, but there's just too many things relying on L2...

**Comment by Someguitarist** [comment]
Sorry, looks like I commented on another issue that's basically the same problem(#2758), but it looks like I'm having the same thing.  No music assistant device appears no matter what I do.  It *seems* like the common thread is that for people running it in docker it doesn't appear as a device no matter what.

Can the people having issues confirm if you're running it as a docker-container instead of an HA addon?

I have confirmed that I'm in network mode host and running the container as privileged, however nothing I do across multiple devices on the same network shows a MUSIC ASSISTANT device.

**Comment by Someguitarist** [comment]
EDIT:  Sorry, figured it out.  Had to disable UFW!  Doh.  Then it popped right up, authenticated with Spotify, then re-enabled UFW.  Rebooted and tested and it seems like it's working.  Thanks!

**Comment by marcelveldt** [comment]
I have some good news. I managed to get librespot compatible with the PKCE oauth flow.
This means that we (finally) can have a true oauth login flow for our Spotify provider and when we need to playback we can just use the retrieved token from that flow for librespot as well. 

For anyone now struggling with authentication (because it relies on librespot) that will be fixed.
I am going to release a beta now with these changes so people can give it a try.

**Comment by marcelveldt** [comment]
The new/final auth method for Spotify is available in 2.2.0 beta 8 - if this is confirmed working, it will also be included in the patch release 2.1.5 for the stable version

**Comment by OzGav** [comment]
This is working well

**Comment by derekngibbs** [comment]
I'm having the same issue. After upgrading to 2.2.0b9 Spotify quit working. I removed it and tried to add it back. I'm entering my client Id in and hit authenticate and it just spins. Any ideas?

**Comment by OzGav** [comment]
@derekngibbs what version of the server did Spotify work for you? It sound like the pop up is being blocked. Have you tried a different browser?

**Comment by derekngibbs** [comment]
I was trying from my phone. I'll try from my laptop and report back

**Comment by eamonnsullivan** [comment]
It was all working well for me yesterday, but this morning I needed to reauthenticate again with Spotify. It's a bit of a pain because it's both mine and my wife's accounts. Is that normal?

**Comment by OzGav** [comment]
The need to re-authenticate is being looked at

**Comment by eamonnsullivan** [comment]
Well, on the plus side, I've now memorised my wife's Spotify login credentials with all of these reauthenticating!

If it makes a difference, we're on a family plan -- we have different logins, playlists, etc., but under the same account.

**Comment by marcelveldt** [comment]
> The need to re-authenticate is being looked at

On b9 it should keep authenticated. If not, I'm out of options.

**Comment by eamonnsullivan** [comment]
I'm on b9, but I'll keep an eye on it. I've logged in today on both accounts. I'll see what happens tomorrow.

**Comment by eamonnsullivan** [comment]
I also see a lot of these, which might be the result of having gone through b7 and b8: 

```
2024-08-18 13:50:34.443 WARNING (MainThread) [music_assistant.music.track] Unable to fetch artist details library://artist/1475 - artist not found in library: 1475
2024-08-18 13:50:34.446 WARNING (MainThread) [music_assistant.music.track] Unable to fetch artist details library://artist/1478 - artist not found in library: 1478
2024-08-18 13:50:34.455 WARNING (MainThread) [music_assistant.music.track] Unable to fetch artist details library://artist/1682 - artist not found in library: 1682
2024-08-18 13:50:34.460 WARNING (MainThread) [music_assistant.music.track] Unable to fetch artist details library://artist/1200 - artist not found in library: 1200
```

And also see: 
```
2024-08-18 13:52:22.546 WARNING (MainThread) [music_assistant.music] Sync task for Spotify: eamonn.sullivan completed with errors
2024-08-18 13:52:24.548 WARNING (MainThread) [music_assistant.music] Sync task for Spotify: Theresa Sullivan completed with errors
```
I may have to bite the bullet and just reinstall.

**Comment by marcelveldt** [comment]
OK, reproduced the expiring token - it was actually a rather stupid typo.
So... beta 10 coming up which keeps the token alive :-)

**Comment by marcelveldt** [comment]
> I may have to bite the bullet and just reinstall.

There is no (more) need for a full reinstall - we have introduced a button to just wipe the library database and trigger a full rescan.

**Comment by eamonnsullivan** [comment]
> There is no (more) need for a full reinstall - we have introduced a button to just wipe the library database and trigger a full rescan.

I might not need this, since b10 seems to be working for me. But for future reference, where is this button?

**Comment by marcelveldt** [comment]
> > There is no (more) need for a full reinstall - we have introduced a button to just wipe the library database and trigger a full rescan.
> 
> I might not need this, since b10 seems to be working for me. But for future reference, where is this button?

settings - core modules - music controller - advanced section

**Comment by eamonnsullivan** [comment]
b10 and the reset button all seems to work fine for me -- looking good!

**Comment by OzGav** [comment]
You need to ENABLE popups. Use a different browser.

**Comment by orgebar** [comment]
> You need to ENABLE popups. Use a different browser.

That was a typo... of course I tried to enable. I tried it in three different browsers plus the app, so this doesnt answer my question.

**Comment by OzGav** [comment]
Then something else on your network is blocking the progress. Have you tried a different computer?

**Comment by marcelveldt** [comment]
> That was a typo... of course I tried to enable. I tried it in three different browsers plus the app, so this doesnt answer my question.

I think it wont work on iOS at all - seems like iOS is actively blocking all popups (or requests for new tabs) by default.
So if you pick a desktop machine it will work. Not sure if there is anything we can do about that, it feels a bit strange that they block this so hard.

**Comment by marcelveldt** [comment]
Found a workaround for iOS, available in b11

**Comment by terierul** [comment]
for me is not working . it remains is this step forever
![image](https://github.com/user-attachments/assets/ee9dba96-e6a7-4462-95a7-9d07577501ff)

tried chrome and MS edge. popup diasabled

**Comment by marcelveldt** [comment]
> for me is not working . it remains is this step forever
> tried chrome and MS edge. popup diasabled

That is because you are still on the old version. The fix we are talking about here is only available in 2.2.0 beta 11.
We will push out a new stable version soon but we want to make sure everything works as expected.

In your case you still have the connect auth, so you need to open the spotify app to authenticate

**Comment by marcelveldt** [comment]
2.2.0 stable has landed (with the -final- new spotify auth flow) - we also released 2.3.0 beta 0 with these fixes included

**Comment by robin-thoni** [comment]
Just tried twice, but it didn't work.

First time, I got a blank page on the MASS callback url saying I successfully logged in, but the provider was not in the settings.

Second time, the pop-up just closed after clicking Agree on the Spotify OAuth flow. Still no music provider.

Will try again tomorrow from desktop (tried on phone).

BTW, the callback URL is the IP address with a port number. Would be really great to use the current Host instead, in case the IP:port would be behind a RP, and not directly accessible.

**Comment by OzGav** [comment]
Confirm you have set the REDIRECT URL as https://music-assistant.io/callback

**Comment by OzGav** [comment]
Solution suggested is to remove the Spotify provider. WAIT until the logs says the removal is complete. Add the provider again.

**Comment by robin-thoni** [comment]
> Solution suggested is to remove the Spotify provider.

I forgot to say I was trying it on a test HA/MASS instance that was never logged-in to any account, so, there's no Spotify provider yet.

Anyway, I just tried it from desktop, and it worked once I pressed `Save` after authentication.

I have a strange HA app behavior on Android where the app just disappears from the recent apps as soon as I leave it (and also most likely gets killed). So, I could not see the `Save` button on mobile after the auth process.

All clear now, thanks for your awesome work guys :+1:

**Comment by OzGav** [comment]
@robin-thoni Good find on the app getting killed.I will add to the docs that you need to come back to MA and hit SAVE for this to work.

**Comment by marcelveldt** [comment]
> BTW, the callback URL is the IP address with a port number. Would be really great to use the current Host instead, in case the IP:port would be behind a RP, and not directly accessible.

You don't have to worry about that, that is sorted out automatically as long as you use "https://music-assistant.io/callback" as callback/redirect URI in the spotify dev console (when using a custom client id).

**Comment by marcelveldt** [comment]
> Anyway, I just tried it from desktop, and it worked once I pressed `Save` after authentication.

Yes you really need to press save for the changes to persist. The UI will also note that but I guess in your flow that was missed.

**Comment by marcelveldt** [comment]
> Solution suggested is to remove the Spotify provider. WAIT until the logs says the removal is complete. Add the provider again.

For now that is an accepted answer but I found the root cause fr the popup to not show on existing installs (only for new additions) - I will quickly patch it and issue a 2.2.1 update.

**Comment by marcelveldt** [comment]
2.2.1 is now now building - should be available within now and 15 minutes

**Comment by WebSpider** [comment]
I just installed 2.2.1, but the callback URL doesn't function when using this as a HomeAssistant Add-On. The redirect-page at music-assistant.io.

The state parameter that's being sent to music-assistant.io/callback in my case is:

https://music-assistant.io/callback/?code=<snipped>&state=http%3A%2F%2F<ip of my HA box>%3A8097%2Fcallback%2Fnt75pbtH4Zx

I tried, and my HA box doesn't listen on port 8097. Possibly the ingress there is causing an issue?

**Comment by marcelveldt** [comment]
> I just installed 2.2.1, but the callback URL doesn't function when using this as a HomeAssistant Add-On. The redirect-page at music-assistant.io.

Yes it is but its an URL that is only available from within your network.
Did you click the link from WITHIN your local network ? You cant do this from remote

**Comment by WebSpider** [comment]
> > I just installed 2.2.1, but the callback URL doesn't function when using this as a HomeAssistant Add-On. The redirect-page at music-assistant.io.
> 
> Yes it is but its an URL that is only available from within your network. Did you click the link from WITHIN your local network ? You cant do this from remote

Thanks for the tip. Apparently I had a VPN running. Stopping that VPN and retrying made it work like a charm!

**Comment by Someguitarist** [comment]
I just wanted to say thank you.  I updated to the newest stable version and it was incredibly easy.  Just clicked Authenticate, clicked I agree, and it worked.  So, so much easier than it was before!

Thanks again!

**Comment by marcelveldt** [comment]
> Thanks for the tip. Apparently I had a VPN running. Stopping that VPN and retrying made it work like a charm!

Ah yeah, that is a good one for the documentation perhaps. Make sure you have no VPN running.
The redirect URL will really only work from within your local network because MA is by default a local-network only service.
We tried making the redirect work with ingress but the url was too long and it didnt accept it.

**Comment by dunxd** [comment]
I set this up a couple of weeks ago and the redirect URL to use then when setting up a client id in developer.spotify.com was the internal URL for my Music Assistant. 

Now it needs to be `https://music-assistant.io/callback`. You can't use the Spotify Developer app that worked before without changing the redirect URI. This should be highlighted in the documentation.

**Comment by wallzero** [comment]
I can sign in and see the popup redirect page but nothing happens after that

**Comment by OzGav** [comment]
You should not post to closed issues as we may not see it and you have provided insufficient information, 

what server version are you on? Do you see the Spotify page to agree to the connection to MA? Are you trying to connect by using a device on the same subnet with out using a VPN, not behind a firewall or reverse proxy or anything similar

**Comment by wallzero** [comment]
I started with 2.2.2 and also tried 2.3.0b6.

I do see the Spotify page and am using my own client ID.

I was using MusicAssistant in a Docker config behind Traefik. It seems I can't configure the base url? I can't even authenticate with Home Assistant - it tries redirecting to the internal Docker IP.

**Comment by OzGav** [comment]
Probably not going to be able to help then as this sounds like a problem with your setup. Have you set the redirect URL correctly in the Spotify dashboard? Otherwise you will have to connect a device to the same subnet as MA so avoid Traefik during this setup.

**Comment by wallzero** [comment]
I set the redirect url to `https://music-assistant.io/callback/`. After I sign in to Spotify and am redirected to that page I see this:

![image](https://github.com/user-attachments/assets/0de705a9-36fd-44df-9ebb-3e17f24efc73)

But am not taken any further. I don't see any console errors at `https://music-assistant.io/callback/`.

I managed to connect to Home Assistant using an authentication token and Music Assistant is able to see my Snapcast server just fine. I don't think using Home Assistant with a reverse proxy like Traefik is that exotic.

Is there a way to set the base url?

![image](https://github.com/user-attachments/assets/3c7e936d-9b77-4091-9112-161f64d07c1d)

Another option could be to manually input the code from the redirect url.
